src/librustc/mir/interpret/value.rs

```diff
@@ -92,277 +86,171 @@ impl<'tcx> ty::TypeFoldable<'tcx> for Value {
-        PrimVal::Ptr(ptr).into()
+impl From<Pointer> for Scalar {
+    fn from(ptr: Pointer) -> Self {
+        Scalar::Ptr(ptr)
     }
 }
 
-/// A `PrimVal` represents an immediate, primitive value existing outside of a
+/// A `Scalar` represents an immediate, primitive value existing outside of a
 /// `memory::Allocation`. It is in many ways like a small chunk of a `Allocation`, up to 8 bytes in
-/// size. Like a range of bytes in an `Allocation`, a `PrimVal` can either represent the raw bytes
-/// of a simple value, a pointer into another `Allocation`, or be undefined.
+/// size. Like a range of bytes in an `Allocation`, a `Scalar` can either represent the raw bytes
+/// of a simple value or a pointer into another `Allocation`
 #[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, RustcEncodable, RustcDecodable, Hash)]
-pub enum PrimVal {
+pub enum Scalar {
     /// The raw bytes of a simple value.
-    Bytes(u128),
+    Bits {
+        /// number of bits that are valid and may be read
+        defined: u8,
+        bits: u128,
+    },
 
     /// A pointer into an `Allocation`. An `Allocation` in the `memory` module has a list of
-    /// relocations, but a `PrimVal` is only large enough to contain one, so we just represent the
-    /// relocation and its associated offset together as a `MemoryPointer` here.
-    Ptr(MemoryPointer),
-
-    /// An undefined `PrimVal`, for representing values that aren't safe to examine, but are safe
-    /// to copy around, just like undefined bytes in an `Allocation`.
-    Undef,
-}
-
-#[derive(Clone, Copy, Debug, PartialEq)]
-pub enum PrimValKind {
-    I8, I16, I32, I64, I128,
-    U8, U16, U32, U64, U128,
-    F32, F64,
-    Ptr, FnPtr,
-    Bool,
-    Char,
+    /// relocations, but a `Scalar` is only large enough to contain one, so we just represent the
+    /// relocation and its associated offset together as a `Pointer` here.
+    Ptr(Pointer),
 }
 
-impl<'tcx> PrimVal {
-    pub fn from_u128(n: u128) -> Self {
-        PrimVal::Bytes(n)
-    }
-
-    pub fn from_i128(n: i128) -> Self {
-        PrimVal::Bytes(n as u128)
+impl<'tcx> Scalar {
+    pub fn undef() -> Self {
+        Scalar::Bits { bits: 0, defined: 0 }
     }
 
     pub fn from_bool(b: bool) -> Self {
-        PrimVal::Bytes(b as u128)
+        // FIXME: can we make defined `1`?
+        Scalar::Bits { bits: b as u128, defined: 8 }
     }
 
     pub fn from_char(c: char) -> Self {
-        PrimVal::Bytes(c as u128)
+        Scalar::Bits { bits: c as u128, defined: 32 }
     }
 
-    pub fn to_bytes(self) -> EvalResult<'tcx, u128> {
+    pub fn to_bits(self, size: Size) -> EvalResult<'tcx, u128> {
         match self {
-            PrimVal::Bytes(b) => Ok(b),
-            PrimVal::Ptr(_) => err!(ReadPointerAsBytes),
-            PrimVal::Undef => err!(ReadUndefBytes),
+            Scalar::Bits { defined: 0, .. } => err!(ReadUndefBytes),
+            Scalar::Bits { bits, defined } if size.bits() <= defined as u64 => Ok(bits),
```

Comment:
Why is even `defined: 0` illegal when the size is `0`? Seems like a strange special case.

Reply:
I think the order here got accidentally flipped.

Reply:
This is not flipped. If you `to_bits` a zst, what `u128` should we be returning? We should probably just `bug!` out here.

Reply:
I'd rather not do that though, because @solson will probably find a way to trigger an ICE by doing horrible things with const fn and unsafe code ;)

Reply:
Oh, this is `to_bits`, not reading a ZST in the first place, nvm.

Reply:
Well, if the first 32 bits are valid we also return the full `u128` when `defined >= 32`.

I don't see why we should not, when the first 0 bits are valid, return the full `u128` with size 0. That's a strange inconsistency.

If this function really can't be called size size 0, it should indeed `bug!` -- I'd rather see ICEs than incorrect CTFE. That ICE would be telling us that some assumption is wrong somewhere.

Reply:
I changed it to `bug!`. You're right, we should fix the callers, not the callee